Refactor index merging, replace Rowboats with RowIterators and RowPointers

leventov opened the pull request:
What this PR does is explained in this issue: #4622.

Key interfaces added in this PR are `TimeAndDimsIterator`, `TimeAndDimsPointer`, `RowIterator` and `RowPointer`. They have very elaborate Javadocs, please read them in order to get the idea of this PR better.

It's performance effect not as good as I expected:
```
BEFORE
Benchmark                    (numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score       Error  Units
IndexMergeBenchmark.mergeV9              5      true             75000     basic  avgt   25  3209755.018 ± 44128.873  us/op
IndexMergeBenchmark.mergeV9              5     false             75000     basic  avgt   25  3123278.657 ± 37920.421  us/op

AFTER
Benchmark                    (numSegments)  (rollup)  (rowsPerSegment)  (schema)  Mode  Cnt        Score       Error  Units
IndexMergeBenchmark.mergeV9              5      true             75000     basic  avgt   25  3302141.747 ± 31396.513  us/op
IndexMergeBenchmark.mergeV9              5     false             75000     basic  avgt   25  3115059.181 ± 36265.165  us/op
```

I. e. it's makes index merging with rollup 3% slower. This is because this PR starts to become really beneficial when garbage-free object metric ser/de is implemented, as mentioned here: #5172, and I wanted to do that before publishing this PR, but it turned out to be a pandora box and it may take several months to complete that work. So I decided to publish this PR even though it doesn't improve performance yet, because this work is already done and i don't want to re-do it when master diverges.

Another reason why this PR doesn't seem so appealing yet, is because the current way of storing data in the incremental index (dims is an array of objects; e. g. double-typed dimension is stored as `Double` object) doesn't penalize the current way of merging indexes (explained in #4622). I. e. it's just moving `Double` object from one place to another, no new objects are created.

However, I'm going to replace the current way of storing data in the incremental index as Object[] with some based on ByteBuffers to reduce memory footprint (and finally remove Aggregators, leaving only BufferAggregators that could work in both realtime and historical). When that is done, with the current way of index merging a lot of new objects will need to be created when index merging is started, although no difference for the new index merging structures, presented in this PR.

leventov commented:
I tagged `Design Review` because I want more people to learn about the new logic of merging via review.

leventov commented:
Tagged `Bug` because before, without dimensions: `{"index":100}` and events with a string dimension which is empty: `{"market":"", "index":100}` were not coalesced, however it seems like it was wrong. Those events could be found in [`druid.sample.json`](https://github.com/druid-io/druid/blob/master/processing/src/test/resources/druid.sample.json). See changes in [`SchemalessTestFullTest`](https://github.com/druid-io/druid/pull/5335/files#diff-80cb1e2c8730a7e100dd7d50e2f57830) and [`SchemalessTestSimpleTest`](https://github.com/druid-io/druid/pull/5335/files#diff-571584283f75d09f84b81506c29252b3).

leventov commented on processing/src/main/java/io/druid/segment/DimensionIndexer.java:
```diff
@@ -257,37 +263,27 @@ DimensionSelector makeDimensionSelector(
-   */
-  EncodedKeyComponentType convertUnsortedEncodedKeyComponentToSortedEncodedKeyComponent(EncodedKeyComponentType key);
-
+  ColumnValueSelector convertUnsortedValuesToSorted(ColumnValueSelector selectorWithUnsortedValues);
```
@jon-wei could you please help to create a good javadoc for this method? I forgot why we need to sort dimension values before index merging.

leventov replied:
Or @gianm?

gianm replied:
I don't remember either. Definitely they must be sorted _after_ merging (dictionaries need to be sorted) but I forget if there is a specific reason why they need to be sorted before merging.

jon-wei replied:
Sorry, just saw this comment!

The `writeDimValueAndSetupDimConversion` step that builds the merged dictionary across indexes and the conversion buffers used for converting per-index dictionary IDs -> merged dictionary IDs builds the conversions from each index's sorted dictionary, so the Rowboat iterable returned by IncrementalIndexAdapter also needs to use the sorted ID space.

`DictionaryMergeIterator` uses a priority queue to merge the per-index dictionary values, so I guess that's why the sorted IDs are used now.

leventov replied:
Now I don't understand why merging is not broken, if it compares `IndexedInts` (in this PR; or `int[]`, before) just based on sorted values, without looking up the original values. E. g.
Index1:
null -> 0,
apple -> 1,
banana -> 2.

Index2:
null -> 0,
watermelon -> 1.

Then it will merge [watermelon] row from the second index before [banana] row from the first index, because it's sorted index is lower.

jon-wei replied:
`writeDimValueAndSetupDimConversion` would look at the actual String values to create an int buffer for each index to be merged, used as a mapping between sorted index-specific IDs -> sorted merged IDs.

Using that example, the dictionary building step would create merged dictionary:
null -> 0
apple -> 1
banana -> 2
watermelon -3

As part of that process, a conversion buffer for each index to be merged would be generated:
Index 1:
0 -> 0 (null)
1 -> 1 (apple)
2 -> 2 (banana)

Index 2:
0 -> 0 (null)
1 -> 3 (watermelon)

These conversions are applied by `MMappedIndexRowIterable` before the rows are fed into the row merging function

leventov replied:
Ok, thanks. I see now. Updated Javadoc comments.

However now as I clearly see the whole picture, I have another question... #5526

leventov commented:
This PR is ready for review. It's going to conflict heavily with #5278 so #5278 should be merged first, but review don't need to wait for that.

leventov commented:
@nishantmonu51 could you please review this?

gianm commented:
> Another reason why this PR doesn't seem so appealing yet, is because the current way of storing data in the incremental index (dims is an array of objects; e. g. double-typed dimension is stored as Double object) doesn't penalize the current way of merging indexes (explained in #4622). I. e. it's just moving Double object from one place to another, no new objects are created.
>
> However, I'm going to replace the current way of storing data in the incremental index as Object[] with some based on ByteBuffers to reduce memory footprint (and finally remove Aggregators, leaving only BufferAggregators that could work in both realtime and historical). When that is done, with the current way of index merging a lot of new objects will need to be created when index merging is started, although no difference for the new index merging structures, presented in this PR.

@leventov - in this PR you write that it may take several months to complete the series of changes you proposed, but you have also recently written elsewhere that you are "probably going to stop being involved in [Druid] development soon". Forgive me for asking, but are you going to be an active contributor long enough to complete this series of contributions and help with follow up? I ask because right now this area of the code is quite stable, and may become less stable as a result of major changes. Improvements are good, but it's always best if the original author is around to fix bugs, help with problems that may arise, comment on why things were done certain ways, and so on.

For example one potential bug could be a performance bug -- I have done experiments with replacing all uses of Aggregator with BufferAggregator before, and ran into problems with timeseries queries slowing down and growable aggregators in IncrementalIndex taking more memory than expected (I don't remember the exact details, but IIRC, some aggregator in datasketches has a buffer version that allocates the max upfront but the non-buffer versions can start small). If you are using the same approach I did in those experiments then your approach might run into the same problems. But even if not, that's just one example -- there are other potential bugs that could occur too.

leventov commented:
> @leventov - in this PR you write that it may take several months to complete the series of changes you proposed, but you have also recently written elsewhere that you are "probably going to stop being involved in [Druid] development soon". Forgive me for asking, but are you going to be an active contributor long enough to complete this series of contributions and help with follow up? I ask because right now this area of the code is quite stable, and may become less stable as a result of major changes.

Amount of work is not so big, I envision that it may take several months because for example it will require to propose some changes to https://github.com/DataSketches/memory, wait them merged (with PR review discussions, etc.), then wait until the library with needed changes is released in Maven central. After then (not in parallel), do the same in https://github.com/RoaringBitmap/RoaringBitmap. After that, some PR(s) in Druid. There might be some delays along the way.

Another reason why it may take several months *is* that I'm going to reduce my involvement in Druid. But it doesn't take several months of full-time work.

> Improvements are good, but it's always best if the original author is around to fix bugs, help with problems that may arise, comment on why things were done certain ways, and so on.

I'm definitely going to be available for such things.

> For example one potential bug could be a performance bug -- I have done experiments with replacing all uses of Aggregator with BufferAggregator before, and ran into problems with timeseries queries slowing down and growable aggregators in IncrementalIndex taking more memory than expected (I don't remember the exact details, but IIRC, some aggregator in datasketches has a buffer version that allocates the max upfront but the non-buffer versions can start small). If you are using the same approach I did in those experiments then your approach might run into the same problems. But even if not, that's just one example -- there are other potential bugs that could occur too.

Thanks for this notice, I didn't think about this, but I now see that it could be a problem, if all dimensions and metrics are required to be stored in a single inflexible (because of the shared base) ByteBuffer. Possible solution could be to allow certain Object metrics to be stored as separate (not shared, hence resizable) ByteBuffers, and when a metric Object requires a resize, it just abandons the former heap buffer and allocates a new, bigger one and stores the data in it. API may look like
```java
interface AggregatorFactory {
  /** true return designates that metrics aggregated with this factory need to be stored
      in separate buffers */
  default boolean needsFlexibleAggregate() { return false; }
}

interface BufferAggregator {
  void aggregate(ByteBuffer buf, int position); // this method already exists

  /** If the aggregated object doesn't fit the given buffer (from pos 0 to capacity),
      this method needs to allocate a new one and return from this method */
  default ByteBuffer aggregateFlexible(ByteBuffer buf) {
    aggregate(buf, 0);
    return buf;
  }
}
```

nishantmonu51 commented:
@leventov I am on vacations for this week, will start reviewing this from next week.

jihoonson commented:
I will start to review in a few days.

leventov commented on processing/src/main/java/io/druid/segment/DoubleDimensionMergerV9.java:
```diff
@@ -24,76 +24,22 @@
+  GenericColumnSerializer setupEncodedValueWriter()
   {
-    return false;
+    return IndexMergerV9.createDoubleColumnSerializer(segmentWriteOutMedium, dimensionName, indexSpec);
```
I don't have a definite opinion. Your point makes sense, but also it makes sense to keep those methods (for Double, Float and Long) together. I would not change this in this PR, because it's not in the scope of this PR. I didn't put those methods in `IndexMergerV9` in this PR.

jihoonson replied:
Ok.

jihoonson commented:
Sorry for delay. I'll try my review by this weekend.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/DoubleDimensionHandler.java:
```diff
@@ -19,17 +19,23 @@
+   */
+  @SuppressWarnings("ComparatorCombinators")
+  private static Comparator<ColumnValueSelector> DOUBLE_COLUMN_COMPARATOR =
+      (s1, s2) -> Double.compare(s1.getDouble(), s2.getDouble());
```
getDouble can throw assertion error for Nulls. need to handle that case here.

leventov replied:
Thanks for this finding, fixed.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/FloatDimensionHandler.java:
```diff
@@ -19,17 +19,18 @@
 public class FloatDimensionHandler implements DimensionHandler<Float, Float, Float>
 {
+  private static Comparator<ColumnValueSelector> FLOAT_COLUMN_COMPARATOR =
+      (s1, s2) -> Float.compare(s1.getFloat(), s2.getFloat());
```
please verify nulls are handled properly.

leventov replied:
Fixed. However, testing this is not the subject of this PR. Please add tests that ensure that columns are merged correctly in the presence of nulls in one of your null-handling PRs.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -305,125 +300,75 @@ public int seek(int dictId)
+
+      convertedMarkedDimensionSelectors[i] = mergers.get(i).convertSegmentRowValuesToMergedRowValues(
+          indexNumber,
+          sourceRowIterator.getMarkedPointer().dimensionSelectors[i]
```
please add some comments to this method to make understanding the flow easier.

leventov replied:
Fixed. However, testing this is not the subject of this PR. Please add tests that ensure that columns are merged correctly in the presence of nulls in one of your null-handling PRs.

leventov replied:
Reorganized code a bit and added comments.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -305,125 +300,75 @@ public int seek(int dictId)
-          }
-        }
+      @Override
+      public TimeAndDimsPointer getMarkedPointer()
```
please add javadocs for getPointer and getMarkedPointer

leventov replied:
Those methods are documented on interface level: see `TimeAndDimsIterator`, `RowIterator`, `TransformableRowIterator`. Let me know if something is unclear from those specifications.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/LongColumnSerializerV2.java:
```diff
@@ -112,13 +111,13 @@ public void open() throws IOException
       writer.add(0L);
     } else {
-      writer.add(((Number) obj).longValue());
+      writer.add(selector.getLong());
```
note to self: reviewed till this point.

jihoonson commented on processing/src/main/java/io/druid/segment/TransformableRowIterator.java:
```diff
@@ -0,0 +1,54 @@
+   *
+   * This method is used in {@link RowCombiningTimeAndDimsIterator} implementation.
+   */
+  TimeAndDimsPointer getMarkedPointer();
```
Why is the return type of getPointer() and getMarkedPointer() different?

leventov replied:
Merely because it's not needed to return a more specific interface from this method. Amended Javadoc.

leventov commented on processing/src/main/java/io/druid/segment/RowIterator.java:
```diff
@@ -0,0 +1,66 @@
+ * RowIterator, because of {@link TimeAndDimsPointer} allocation-free design, that reuses objects. On the other hand,
+ * some implementations of RowIterator allow to optimize {@link #mark()} and {@link #hasTimeAndDimsChangedSinceMark()}.
+ */
+public interface RowIterator extends TimeAndDimsIterator
```
Interesting observation, I didn't notice this fact before. However I'm not sure that collapsing those interfaces is worthwhile. Left some comments in code.

leventov commented:
@nishantmonu51 do you have more comments on this PR?

clintropolis commented on processing/src/main/java/io/druid/segment/TimeAndDimsPointer.java:
```diff
@@ -0,0 +1,213 @@
+ * of only time point and dimension and metric values, not tied to a specific position in any data structure (aka "row
+ * index").
+ *
+ * TimeAndDimsPointer intrface is conceptually similar to {@link Cursor}, that is used for query execution rather than
```
nit, typo 'intrface' (but seriously, :+1: on all these javadocs and comments)

leventov replied:
Thanks. Additionally, TimeAndDimsPointer is not an interface, it's a class. Fixed comments.

clintropolis commented:
I've had a couple of passes through this now, LGTM :+1: (though you'll probably want to pull latest and undo this fix https://github.com/druid-io/druid/pull/5510/files just wasn't sure when this PR will be merged)

fjy commented:
I strongly think we need to be really careful with a large refactor of the ingestion code that doesn't seem to add much of a performance improvement and carefully test that it doesn't break anything. My vote is to not merge this code unless we are confident about that.

leventov commented:
@fjy this PR actually fixes a bug in index merging code.

leventov commented:
@fjy I feel that current unit and integration test suite tests index merging well. It wasn't that I wrote some code and it started to pass all tests immediately, far from that. Actually I was surprised how subtle corner cases were revealed (and I had to fix my new code) during unit testing.

This PR doesn't improve performance directly, but it's a part of a grand plan, that will allow to get rid of Aggregator interface, for instance (with the current merging scheme, it won't be possible to do that reasonably efficiently). I explained that in details above in this PR comments and in #5335.

And on top of that, this PR *fixes a bug*.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/MergingRowIterator.java:
```diff
@@ -0,0 +1,273 @@
+ * In other words, MergingRowIterator is an equivalent to {@link com.google.common.collect.Iterators#mergeSorted}, but
+ * for {@link RowIterator}s rather than simple {@link java.util.Iterator}s.
+ *
+ * Implementation detail: this class uses binary heap priority queue algorithm to sort pointers, but it also momoizes
```
s/momoizes/memoizes/g

leventov replied:
Fixed

nishantmonu51 commented on processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java:
```diff
@@ -176,130 +171,200 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+
+  /**
+   * On {@link #moveToNext()} and {@link #mark()}, this class copies all column values into a set of {@link
+   * SettableColumnValueSelector} instances. Alternative approach was to save only offset in column and use they same
```
s/they same/the same/g

leventov replied:
Fixed

nishantmonu51 commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndexRowHolder.java:
```diff
@@ -0,0 +1,66 @@
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    // nothing to inspect
+  }
```
note to self : reviewed till here.

nishantmonu51 commented:
Hi @leventov, please summarize the changes in the PR description, especially around the new concepts/interfaces introduced and their purpose to would make the review easy. 
 
Also, can you list down all the future changes that are needed to make this PR useful and get some good results ? 

Finally, is this code tested somewhere in production ? If not, can we get branch tested in some test cluster to gain some more confidence here.

leventov commented:
> please summarize the changes in the PR description, especially around the new concepts/interfaces introduced and their purpose to would make the review easy.

I've added a second paragraph in the first message in this PR. You should basically read javadocs of 4 added classes and interfaces. I don't see a point in copy-pasting Javadoc here.

> can you list down all the future changes that are needed to make this PR useful and get some good results ?

There is ObjectStrategy refactoring, see the first message in this thread:
> I. e. it's makes index merging with rollup 3% slower. This is because this PR starts to become really beneficial when garbage-free object metric ser/de is implemented, as mentioned here: #5172, and I wanted to do that before publishing this PR, but it turned out to be a pandora box and it may take several months to complete that work. So I decided to publish this PR even though it doesn't improve performance yet, because this work is already done and i don't want to re-do it when master diverges.

Some more info: https://github.com/DataSketches/memory/issues/14#issuecomment-353591816. It's already 70% done (for several months already), it awaits for this PR and also for some changes in DataSketches/memory. It will allow to drastically reduce amount of garbage produced when Object columns are involved into index merging (like histogram, dataSketches).

> Finally, is this code tested somewhere in production ? If not, can we get branch tested in some test cluster to gain some more confidence here.

No, it's not used in production. I'm not going to test this change in production before it's part of a release, because we moved away from this practice. Also see https://github.com/druid-io/druid/pull/5335#issuecomment-375743034

leventov commented:
@nishantmonu51 do you have more comments?

nishantmonu51 commented:
LGTM. 👍

nishantmonu51 commented:
@fjy: do you still have concerns in merging this PR ?

fjy commented:
👍

leventov commented:
Thanks for reviews.
